### ISSUE TITLE
Documentation for returning functionality in MySQL is not clear enough in all cases

### ISSUE THREAD

**Issue by jsumners** [tgriesser/knex#3158]
# Environment
Knex version: 0.16.5
Database + version: MySQL 5.7

# Bug

```js
db.insert({data}, 'an_id').into('foobar')
```

That sort of statement is giving `returning() is not supported by mysql and will not have any effect.` Note that this statement is _not_ using `returning()` but is relying on Knex's internal version of it for MySQL.

**Comment by chaffeqa** [comment]
Can confirm, due to this change obv: https://github.com/tgriesser/knex/pull/3039

**Comment by kibertoad** [comment]
@jsumners Why are you passing second param to `insert`? It is literally a `returning` argument.

**Comment by kibertoad** [comment]
All it does is `if (!isEmpty(returning)) this.returning(returning);`, so it's pretty much same thing.

**Comment by chaffeqa** [comment]
According to what I understand, that is the only way to get a result from an `insert` in mysql dialect (since `returning(...)` syntax doesnt work?)

**Comment by kibertoad** [comment]
No, it's exactly same thing and is unlikely to have any effect either, warning is correct.

**Comment by chaffeqa** [comment]
oh intiresting... well it works for mysql then... whats the warning for?

**Comment by kibertoad** [comment]
Can you try removing it and see if that makes any difference?

**Comment by chaffeqa** [comment]
```> await db.connection.default("User").insert({})
[2019-04-29 19:47:07.346 +0000] INFO  (base/36985 on Quinns-MacBook-Pro-2.local): SQL: insert into `User` () values () , []
[2019-04-29 19:47:07.351 +0000] INFO  (base/36985 on Quinns-MacBook-Pro-2.local): Time: 49.751 ms

[ 202 ]
```

Oh wow I guess not... so the default is to return an array of `id`s inserted?

**Comment by jsumners** [comment]
The documentation promotes this syntax:

![image](https://user-images.githubusercontent.com/321201/56922742-f1731300-6a96-11e9-9aa5-69f955adea29.png)

**Comment by kibertoad** [comment]
If what I'm seeing in builder tests is correct, unfortunately no, just an array of last id inserted. Likely because LAST_INSERT_ID() is used somewhere (can't find reference in knex, could be mysql driver thing).
MySQL is not a great DB.

**Comment by kibertoad** [comment]
@jsumners Documentation is not MySQL-specific. Also it says: " It's a shortcut for returning method". And documentation for `returning` method says this: "Utilized by PostgreSQL, MSSQL, and Oracle databases".

Not a bug, works as designed.

**Comment by jsumners** [comment]
Wow. 

1. Yes, the documentation _is_ MySQL specific. I highlighted the dropdown that indicates it is so.
2. It shouldn't matter what the documentation behind the "returning method" says because this documentation is clearly indicating that Knex will do the right thing for MySQL when given this syntax.

**Comment by kibertoad** [comment]
@jsumners OK, documentation is misleading here and needs to be clarified. In MySQL case it is actually likely to return 3 and not 2, and also regardless of returning param. This is definitely not communicated properly.

**Comment by kibertoad** [comment]
`documentation is clearly indicating that Knex will do the right thing for MySQL when given this syntax.` - not really. Fact that it only returns you single ID on MySQL is already a strong hint that it's not doing the right thing, but yes, that should have been explained in much less ambiguous terms.

**Comment by chaffeqa** [comment]
ugh I feel for you guys, yea very difficult to explain, but thanks for the clarification!

FYI for the curious, the implementation looks to be in https://github.com/mysqljs/mysql#getting-the-id-of-an-inserted-row (though a quick search didnt help me understand the exact spot).

Along those lines, it *looks* like this is race condition safe (which was my immediate next concern)

I miss postgres 😢

**Comment by kibertoad** [comment]
@chaffeqa Thanks for investigating this further!
Totally share your feelings towards PostgreSQL 😂

**Comment by jsumners** [comment]
> I miss postgres 😢

You and me both.

**Comment by elhigu** [comment]
Afaik that dropdown in upper corner only selects dialect that is used to generate output from example builders... so rest of the docs are static and the same for all the dialects (and that is intentional to highlight for everyone those parts of docs where builders are not generation cross db compatible queries). If anyone is up to make this part more clear in docs please do.

This issue could be also closed and moved to docs repo.

**Comment by jsumners** [comment]
@chaffeqa it looks like this is where the returning in happening

https://github.com/tgriesser/knex/blob/a2ab754eedcc1d8fe6328cb40adf0e31b2b3576a/src/dialects/mysql/index.js#L153-L154

But I do not see where it is appending the `set ?`.

**Comment by ziggi** [comment]
Can we got an ability for disable this warning? Because I use knex with mysql and with `returning` feature as well. And returning only `first` inserted id is enough for me.

By the way, why we can just fix the mysql knex dialect with changes below?

Just change this line:
https://github.com/tgriesser/knex/blob/a2ab754eedcc1d8fe6328cb40adf0e31b2b3576a/src/dialects/mysql/index.js#L154
to this:
```JavaScript
return Array.from({ length: rows.affectedRows }, (v, k) => k + rows.insertId);
```

**Comment by elhigu** [comment]
@ziggi that warning is there for a reason. You should not use `.returning()` with mysql. It does not do anything . There is a separate issue for adding crossdb compatible insert to knex https://github.com/tgriesser/knex/issues/2682 ... that could be place to talk about that change. I have to say that I don't remember reasons why knex was never changed to generate those "missing" ids (anyways this thread is not the place to discuss about it).

**Comment by ziggi** [comment]
> It does not do anything . 

But it's do, with `.returning()` I can receive first inserted id.

**Comment by elhigu** [comment]
@ziggi you will get that id also without adding any ‘.returning()’ call there. Check from ‘.insert()’ documentation what it returns by default... as I said, ‘.returning()’ is not supported at all but mysql server and now knex emits a warning because you are trying to use it with wrong dialect.

**Comment by ziggi** [comment]
@elhigu ohh, thanks, my bad.

**Comment by rijkvanzanten** [comment]
@elhigu is there any way to suppress that warning? My app is database driver agnostic, and so far, inserts have been the only painpoint where I haven't been able to properly make it database driver agnostic :( 

MySQL returns primary keys (but only if they're auto incremented it seems), but others (Postgres) returns a "Result" object coming from Knex. My solution was to do something like

```
const keys = await knex.insert({ ... }).into(...).returning('id');

let key;

if (typeof keys[0] === number) {
  key = keys[0];
} else {
  key = keys[0].id;
}
```

but that keeps throwing warnings about MySQL not supporting returning

**Comment by elhigu** [comment]
@rijkvanzanten probably the easiest and fastest way for you is to write knex plugin, which implements crossdb insert that works in your case.

Something like:

```
const Knex = require('knex');
Knex.QueryBuilder.extend('insertAndReturnId', function(value) {
  const builder = this.insert(value);
  if (this.client.driverName === 'pg' || this.client.driverName === 'oracledb') {
    builder.returning('id');
  }
  return builder;
});

const id= await knex('accounts').insertAndeturnId({});
```

**Comment by williamdes** [comment]
https://github.com/knex/knex/search?q=.returning%28%29+is+not+supported+by+mysql+and+will+not+have+any+effect.

Here is my migration, I want it to run silently.
Please if someone can make this warning shut up :pray: 
```ts
'use strict';

import * as Knex from 'knex';

exports.up = function (knex: Knex) {
  return knex
    .insert([
      {
        idType: 1,
        nomType: 'Midi',
      },
      {
        idType: 2,
        nomType: 'Soir',
      },
    ])
    .into('Types');
};

exports.down = function (knex: Knex) {
  return knex.delete('*').from('Types');
};
```

```
Using environment: test
  knex:tx trx2: Starting top level transaction +0ms
  knex:tx trx2: releasing connection +104ms
  knex:tx trx3: Starting top level transaction +139ms
.returning() is not supported by mysql and will not have any effect.
  knex:tx trx3: releasing connection +149ms
Batch 2 rolled back the following migrations:
20181110160827-data-types.ts
```

### The patch/hack

```ts
  development: {
    [...]
    log: {
      warn: (message) => {
        // Shut up https://github.com/knex/knex/issues/3158
        if (message === '.returning() is not supported by mysql and will not have any effect.') {
          return;
        }
        console.warn(message);
      },
    },
  },
```

**Comment by elhigu** [comment]
@williamdes That migration should not emit that message, since you have not called `.returning()`. Please open separate bug report if that keeps happening. Sounds like, migration system might be internally using `.returning()` which triggers the error.

**Comment by VictorLessa** [comment]
Good morning guys I need a way to remove the warnning logs, how do I do this?

**Comment by corysimmons** [comment]
**Tip:** If you're mutating a record in the database, and you have the input data, just return that directly upon a successful mutation.

ExpressJS example:

```js
app.post('/api/v1/users', async (req, res, next) => {
  const someBookData = req.body

  try {
    const arrayOfPrimaryKeysSuccessfullyInserted = await knex('books').insert(someBookData) // So this could be like [40222] where 40222 is the PK value in your users table
    res.status(200).send({
      id: arrayOfPrimaryKeysSuccessfullyInserted[0], // [40222][0] = 40222
      ...someBookData // the original book data
    })
  } catch(err) {
    next(err)
  }
})
```

I ***think*** this is correct, but someone feel free to correct me.

**Comment by mercmobily** [comment]
The discussion is quite intense.

The consensus seems to have landed on the fact that this is a documentation matter. On MySQL/MariaDB, `returning()` is not supported by the engine. Knex historically warns when `.returning()` (or the second `insert` parameter, which is a shortcut to `returning`) is used on MySQL, because only a single auto‑increment id is available via driver APIs. The correct guidance is:

- Do not use `.returning()` on MySQL/MariaDB; it has no effect and triggers a warning.
- `insert()` resolves to an array of inserted ids on MySQL (for single‑row inserts, `[insertId]`).

Code path (current): the MySQL compiler warns in its constructor when `returning` is present (`lib/dialects/mysql/query/mysql-querycompiler.js:16–23`).

Action items:

- Update documentation to spell out per‑dialect `insert()` and `returning()` behaviour, including MySQL’s “array of ids” default and the warning.
- Consider providing a cross‑dialect helper in docs (or a small plugin) for “insert and return id(s)” that hides the differences, as suggested by maintainers.

I nearly marked this as duplicate of #6254. However, realistically speaking, that is a much more complex issue, and I would think that this one should be resolved easily, and before that one.

**Comment by rijkvanzanten** [comment]
> Do not use .returning() on MySQL/MariaDB; it has no effect and triggers a warning.

This is distant future potential, but I'm wondering if there's a way to dynamically register the `.returning()` method only if it's available in the client you've chosen to use, so we can rely on typescript typing instead of a runtime warning. In pseudo:

```js
// before

const db = require('knex')({
  client: 'mysql',
  // ...
});

db.returning() // warning;
```

after:

```ts
import { knex } from 'knex';
import { mysql } from '@knex/mysql';

const db = knex({
  client: mysql,
});

db.returning();
// TS: ^ Property 'returning' does not exist on type 'X'.(2339)
```

**Comment by mercmobily** [comment]
@rijkvanzanten This would be absolutely awesome. I think it would be a must-have for knex 4

**Comment by corysimmons** [comment]
You don't have to wait on Knex to do this. Just add your own typing layer and post it here for others to copy.

Just pasting AI slop so be sure to test this but something like:

```ts
// dialect-aware-knex.ts
import type { Knex as K } from 'knex';
import knex from 'knex';

// 1) Dialect tags you care about
type Dialect = 'pg' | 'mssql' | 'oracledb' | 'sqlite3' | 'mysql' | 'mariadb';

// 2) Helper: strip `returning` from a type
type OmitReturning<T> = Omit<T, 'returning'> & { returning?: never };

// 3) Dialects that truly support `.returning()` in knex today
type HasReturning<D extends Dialect> =
  D extends 'pg' | 'mssql' | 'oracledb' | 'sqlite3' ? true : false;

// 4) Wrap the QueryBuilder so `returning` is conditional
type QB<D extends Dialect, TRecord = any, TResult = unknown> =
  HasReturning<D> extends true
    ? K.QueryBuilder<TRecord, TResult>
    : OmitReturning<K.QueryBuilder<TRecord, TResult>>;

// 5) Wrap the main Knex instance so builder methods return our QB
type KnexTyped<D extends Dialect> =
  Omit<K, 'queryBuilder' | 'fn' | 'raw' | 'schema' | 'with' | 'batchInsert'> & {
    <TRecord = any, TResult = any>(tableName: K.TableDescriptor): QB<D, TRecord, TResult>;
    queryBuilder<TRecord = any, TResult = any>(): QB<D, TRecord, TResult>;
    // (optional) add any other members you rely on, returning QB<D,...> where appropriate
  } & K; // keep everything else compatible

// 6) Factory that narrows the type based on a dialect tag you pass
export function createKnex<D extends Dialect>(
  dialect: D,
  config: K.Config
): KnexTyped<D> {
  return knex(config) as unknown as KnexTyped<D>;
}
```

```ts
import { createKnex } from './dialect-aware-knex';

// PG: `.returning` allowed (typed)
const pg = createKnex('pg', { client: 'pg', connection: process.env.PG_URL! });
await pg('users').insert({ name: 'A' }).returning(['id']); // OK

// MySQL: `.returning` is a TS error
const mysql = createKnex('mysql', { client: 'mysql2', connection: process.env.MYSQL_URL! });
await mysql('users').insert({ name: 'A' }).returning(['id']); 
//    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// TS2339: Property 'returning' does not exist on type '...'
```

**Comment by rijkvanzanten** [comment]
Sure, but by that logic you wouldn't have to use any library for anything 😄 Good workaround for now, but it is something that knex should be correctly typed for out of the box imo

**Comment by mercmobily** [comment]
@rijkvanzanten I agree.
